Introduce ASE-style `FIRE` optimizer (departing from velocity Verlet in orig FIRE paper) and improve coverage in `test_optimizers.py`

mstapelberg opened the pull request:
This PR adds a new `ase_fire` flavor to the FIRE optimizers, mimicking ASE's implementation, alongside the existing `vv_fire` (Velocity-Verlet based). Users can now select the desired behavior via the `md_flavor` argument in `fire`, `unit_cell_fire`, and `frechet_cell_fire`. **Breaking: The default changed from `vv_fire` to `ase_fire`.**

Highlights:
- `ase_fire` supports `max_step` (unlike original `vv_fire`)
- new tests for both flavors, including negative power branches, batch/fixed-cell consistency, and more edge cases. Existing FIRE tests are now parameterized by `md_flavor`.
- new example `examples/scripts/7_Others/7.6_Compare_ASE_to_VV_FIRE.py` compares `ase_fire` and `vv_fire`
- bump `setup-uv` GitHub Action to v6, cleanup `pytest` config in `pyproject.toml`

CompRhys commented on torch_sim/optimizers.py:
```diff
@@ -601,89 +587,113 @@ def fire_step(
-        optimizing atomic positions in a batched setting. Uses velocity Verlet
-        integration with adaptive velocity mixing.
+        optimizing atomic positions in a batched setting. Logic adapted to follow
+        ASE FIRE implementation more closely.
```
I believe that @abhijeetgangan chose to follow the literature implementation intentionally. We've been talking about this internally should offer an implementation that matches ASE but there's a question about how best to do that without duplicating code.

mstapelberg replied:
Understood! In my tests with NEB, I found the ASE FIRE implementation to converge quicker for some reason though. Maybe it's unique to NEB, I can test FireState vs this ASEFireState on a few other structures?

janosh replied:
> I can test `FireState` vs this `ASEFireState` on a few other structures?

that would be much appreciated!

janosh commented:
it appears with the changes in this PR, the relaxation @AdeeshKolluru tested in #146 still takes longer to reduce max force than ASE: goes below 1 eV/A in step 17 instead of step 3 for ASE

```
torch-sim optimize
step=1 energy=-70.0 eV  max_force=8.86 eV/A     volume=112.347 A^3
step=2 energy=-70.2 eV  max_force=7.97 eV/A     volume=112.349 A^3
step=3 energy=-70.4 eV  max_force=6.66 eV/A     volume=112.353 A^3
step=4 energy=-70.6 eV  max_force=5.16 eV/A     volume=112.358 A^3
step=5 energy=-70.8 eV  max_force=3.65 eV/A     volume=112.364 A^3
torch-sim optimize
step=6 energy=-70.9 eV  max_force=2.34 eV/A     volume=112.371 A^3
step=7 energy=-71.0 eV  max_force=1.48 eV/A     volume=112.380 A^3
step=8 energy=-71.0 eV  max_force=2.44 eV/A     volume=112.391 A^3
step=9 energy=-71.0 eV  max_force=2.42 eV/A     volume=112.391 A^3
step=10 energy=-71.0 eV max_force=2.35 eV/A     volume=112.392 A^3
torch-sim optimize
step=11 energy=-71.0 eV max_force=2.25 eV/A     volume=112.392 A^3
step=12 energy=-71.0 eV max_force=2.1 eV/A      volume=112.393 A^3
step=13 energy=-71.0 eV max_force=1.93 eV/A     volume=112.395 A^3
step=14 energy=-71.0 eV max_force=1.72 eV/A     volume=112.396 A^3
step=15 energy=-71.0 eV max_force=1.46 eV/A     volume=112.398 A^3
torch-sim optimize
step=16 energy=-71.0 eV max_force=1.15 eV/A     volume=112.401 A^3
step=17 energy=-71.0 eV max_force=0.782 eV/A    volume=112.404 A^3
step=18 energy=-71.0 eV max_force=0.545 eV/A    volume=112.408 A^3
step=19 energy=-71.0 eV max_force=0.412 eV/A    volume=112.414 A^3

```
```
ASE optimize

      Step     Time          Energy          fmax
FIRE:    0 11:08:23      -69.999269        9.183465
FIRE:    1 11:08:23      -70.941082        1.783544
FIRE:    2 11:08:23      -70.972607        1.473498
FIRE:    3 11:08:23      -71.014896        0.960106
FIRE:    4 11:08:23      -71.042696        0.525267
FIRE:    5 11:08:23      -71.047455        0.817320
FIRE:    6 11:08:23      -71.049077        0.775535
FIRE:    7 11:08:24      -71.052065        0.694864
FIRE:    8 11:08:24      -71.055961        0.581221
FIRE:    9 11:08:24      -71.060210        0.444012
FIRE:   10 11:08:24      -71.064289        0.393639
FIRE:   11 11:08:24      -71.067836        0.374165
FIRE:   12 11:08:24      -71.070756        0.356804
FIRE:   13 11:08:24      -71.073478        0.339223
FIRE:   14 11:08:24      -71.076349        0.380043
FIRE:   15 11:08:24      -71.079865        0.389464
FIRE:   16 11:08:24      -71.084310        0.347161
FIRE:   17 11:08:24      -71.089341        0.249172
FIRE:   18 11:08:24      -71.093816        0.162485
FIRE:   19 11:08:24      -71.096501        0.192828
FIRE:   20 11:08:25      -71.097859        0.309374
```

CompRhys commented on torch_sim/optimizers.py:
```diff
@@ -584,90 +586,124 @@ def fire_init(
-        # )
+        # Get atom-wise masks
+        downhill_mask_atoms = downhill_mask_batch[state.batch]
+        _uphill_mask_atoms = uphill_mask_batch[state.batch]
```
`_uphill_mask_atoms` is unused

mstapelberg commented:
Would providing an ASEFireState class be a possible solution? Depending on what the user wants, they can select fire or ase-fire as their optimizer? I have a refactored version of. the ASE FIRE implementation while leaving the closer to paper implementation alone that I can push here.

janosh commented:
> The ASEFireState I've written seems to converge as fast as ASE in NEB at least as shown here: https://github.com/Radical-AI/torch-sim/issues/90#issuecomment-2836069061

that's great to hear! [but as you suggest](https://github.com/Radical-AI/torch-sim/pull/174#discussion_r2072431410), we need benchmarks across multiple to gain some confidence that the ASE implementation is consistently faster to converge

> Would providing an ASEFireState class be a possible solution? Depending on what the user wants, they can select fire or ase-fire as their optimizer?

i think to minimize API surface and make it easier to see how the ASE and original FIRE implementations differ, adding a keyword to the existing `fire` would be better. could call it `md_flavor`:

```diff
    Args:
        model (torch.nn.Module): Model that computes energies, forces, and stress
        dt_max (float): Maximum allowed timestep
        dt_start (float): Initial timestep
        n_min (int): Minimum steps before timestep increase
        f_inc (float): Factor for timestep increase when power is positive
        f_dec (float): Factor for timestep decrease when power is negative
        alpha_start (float): Initial velocity mixing parameter
        f_alpha (float): Factor for mixing parameter decrease
        maxstep (float): Maximum distance an atom can move per iteration (default
            value is 0.2). Only used when md_flavor='ase'.
+        md_flavor ('velocity_verlet' | 'ase'): The type of molecular dynamics flavor to run. Options are
+            'velocity_verlet' (default) or 'ase' (mimics ASE's FIRE implementation which 
+             differs from the original paper).
```

mstapelberg commented:
Sounds good! Will work on this today/tomorrow.

mstapelberg commented:
I just finished modifying optimizer.py and added a test case / example in 

Here's the output, seems like the ASE implementation is slightly better overall! 

```bash
\n--- Running optimization with MD Flavor: ase_fire ---
Step 50: Active structures: 3, Total converged: 0/5
Step 70: Converged indices [2]. Total converged: 1/5
Step 80: Converged indices [1]. Total converged: 2/5
Step 100: Converged indices [0]. Total converged: 3/5
Step 100: Active structures: 1, Total converged: 3/5
Step 150: Active structures: 2, Total converged: 3/5
Step 180: Converged indices [4]. Total converged: 4/5
Step 200: Active structures: 1, Total converged: 4/5
Step 250: Active structures: 1, Total converged: 4/5
Step 300: Active structures: 1, Total converged: 4/5
Step 310: Converged indices [3]. Total converged: 5/5
All structures converged or max iterations reached by batcher.
Finished ase_fire in 31.12 seconds.
\n--- Running optimization with MD Flavor: vv_fire ---
Step 50: Active structures: 3, Total converged: 0/5
Step 80: Converged indices [2]. Total converged: 1/5
Step 90: Converged indices [1]. Total converged: 2/5
Step 100: Active structures: 1, Total converged: 2/5
Step 130: Converged indices [0]. Total converged: 3/5
Step 150: Active structures: 2, Total converged: 3/5
Step 200: Active structures: 2, Total converged: 3/5
Step 230: Converged indices [4]. Total converged: 4/5
Step 250: Active structures: 1, Total converged: 4/5
Step 300: Active structures: 1, Total converged: 4/5
Step 350: Active structures: 1, Total converged: 4/5
Step 400: Active structures: 1, Total converged: 4/5
Step 450: Active structures: 1, Total converged: 4/5
Step 470: Converged indices [3]. Total converged: 5/5
All structures converged or max iterations reached by batcher.
Finished vv_fire in 43.49 seconds.
\n--- Comparison ---
Force tolerance: 0.05 eV/Å
Initial energies: ['-1546.829', '-375.742', '-899.794', '-140.563', '-240.803'] eV
Final ASE energies: ['-2705.613', '-497.718', '-1053.351', '-2692.331', '-487.226'] eV
Final VV energies:  ['-2705.623', '-497.717', '-1053.351', '-2632.801', '-487.219'] eV
Mean Disp (ASE-VV): ['0.0158', '0.0101', '0.0023', '0.9886', '0.0263'] Å
Convergence steps (ASE FIRE): [100, 80, 70, 310, 180]
Convergence steps (VV FIRE):  [130, 90, 80, 470, 230]
```

As an aside, I think the bigger differences in the NEB (See #90 ) are due to my batched NEB implementation assigning each image its own FIRE optimizer parameters. Whereas in the ASE version of fire I had used scalar values of timestep, a (mixing parameter), and Nsteps that was used for the entire trajectory (which I think makes more sense to do). 

I've used Ruff on the optimizer.py. If you guys think this is fine, I can implement the same md_flavor in the UnitCell and FrechetCell versions of FIRE?

janosh commented:
@mstapelberg this is great! 👍 thanks a lot

i only managed a quick glance. i'll take a closer look tomorrow. porting the same behavior to `UnitCell` and `FrechetCell` sounds good! we'll definitely need unit tests for the new `md_flavor` kwarg and your `examples/scripts/7_Others/7.6_Compare_ASE_to_VV_FIRE.py` is a great start for that. i can probably take care of writing the tests tomorrow (unless you'd like to have a go)

mstapelberg commented:
> @mstapelberg this is great! 👍 thanks a lot
> 
> i only managed a quick glance. i'll take a closer look tomorrow. porting the same behavior to `UnitCell` and `FrechetCell` sounds good! we'll definitely need unit tests for the new `md_flavor` kwarg and your `examples/scripts/7_Others/7.6_Compare_ASE_to_VV_FIRE.py` is a great start for that. i can probably take care of writing the tests tomorrow (unless you'd like to have a go)

You bet! Will work to port the same behavior to UnitCell and FrechetCell, while doing the same comparisons. 

Should I adopt/add-on to the test_optimizers.py for this? I notice a lot of the tests there are much shorter/simpler, and don't require a GPU. First time writing proper test cases like this, so happy to give it a go.

janosh commented:
> Should I adopt/add-on to the test_optimizers.py for this?

exactly

> First time writing proper test cases like this, so happy to give it a go.

usually straightforward to port code that works on GPUs to CPUs for unit tests. common strategy for fast tests is to take whatever code you were developing/debugging with and reduce the number of steps and system sizes to as small as possible while still testing something complex enough to cover many edge cases

mstapelberg commented:
@janosh 

Apologies for the delay on this. I have updated the optimizers.py module to have md_flavor selection for all three types of FIRE. 

I added test_cases for `test_fire_optimization`, `test_unit_cell_fire_optimization`, and `test_frechet_cell_fire_optimization`.

The "ase-fire" flavor consistently converges in 1/3 the number of steps.

mstapelberg commented on torch_sim/optimizers.py:
```diff
@@ -489,8 +489,10 @@ def fire(
     alpha_start: float = 0.1,
     f_alpha: float = 0.99,
+    maxstep: float = 0.2,
+    md_flavor: MdFlavor = vv_fire_key,
```
Done!

janosh commented:
1 test started failing. looks like there's a bug in `frechet_cell_fire` with `ase_fire` flavor which got surfaced by changing the default `md_flavor`

janosh reviewed (approved):
thanks a lot @mstapelberg for the massive amount of work that went into this PR! much appreciated! 👍 

and apologies for the slow review

mstapelberg commented:
> thanks a lot @mstapelberg for the massive amount of work that went into this PR! much appreciated! 👍
> 
> and apologies for the slow review

Hi @janosh no worries and sorry for the slow reply on my end (was on travel)! Happy to help with this; and excited to use it for my research as well. Thanks for building this great tool!

janosh commented:
if you have time, maybe you'd like to give #199 a try to see if it improves your FIRE relaxation results

mstapelberg commented:
> if you have time, maybe you'd like to give #199 a try to see if it improves your FIRE relaxation results

Sure! Will do this later tonight/tomorrow morning and will post the results here? Or should I post them on #199 ?

janosh commented:
over at https://github.com/Radical-AI/torch-sim/pull/199 would be great or in a new issue if you find any more problems
